Add Books

sher009 opened the pull request:
Created class Book
Created class BookController
Created interface BookRepository
Created class BookService
Created test BookServiceTest
add dbchangelog_2.xml

sher009 committed:
Created class Book
Created class BookController
Created interface BookRepository
Created class BookService
Created test BookServiceTest
add dbchangelog_2.xml

kostyaskrun commented on src/main/resources/db/db.changelog-master.xml:
```diff
@@ -5,9 +5,8 @@
             http://www.liquibase.org/xml/ns/dbchangelog/dbchangelog-3.8.xsd">
 
-    <!-- Include individual changelogs here -->
+    <!-- Підключаємо файли змін -->
```
remove Ukrainian comment

kostyaskrun commented on src/main/java/com/example/bookstore/model/Book.java:
```diff
@@ -0,0 +1,70 @@
+    private String title;
+
+    @Column(name = "published_date")
+    private LocalDate publishedDate;
```
publishedDate cannot be future date, there is no validation

kostyaskrun replied:
also do not forget about test,
also ideally you should think how you can handle code duplication, because similar validation used in Author birthDate field

kostyaskrun reviewed (changes requested):
Also create integration test to cover next logic:

"Check what happens to books when Author is deleted" (hint: it should be deleted together with author)"

sher009 committed:
changed publishedDate
add Validation to Author and Book
created new tests

kostyaskrun commented on src/main/java/com/example/bookstore/model/Author.java:
```diff
@@ -13,21 +16,22 @@ public class Author {
     private Long id;
 
-    @Column(name = "first_name", nullable = false)
+    @NotNull(message = "First name cannot be null")
```
why have you removed @Column annotation @sher009 ?

sher009 replied:
to add back?
Normaly I could keep bouth

kostyaskrun commented on src/main/java/com/example/bookstore/model/Author.java:
```diff
@@ -13,21 +16,22 @@ public class Author {
+    @NotNull(message = "First name cannot be null")
     private String firstName;
 
-    @Column(name = "last_name", nullable = false)
```
same

kostyaskrun commented on src/main/java/com/example/bookstore/model/Author.java:
```diff
@@ -13,21 +16,22 @@ public class Author {
-    @Column(name = "country")
-    private String country;
 
-    @Column(name = "language")
```
same

kostyaskrun commented on src/test/java/com/example/bookstore/BookValidationTest.java:
```diff
@@ -0,0 +1,24 @@
+public class BookValidationTest {
+
+    @Test
+    void shouldNotAllowFuturePublishedDate(){
```
Will not give you direct solution for this, but will give you a hint instead:

imagine we want to add Child of Author in future. With this code implementation you will create one more test covering birthDate of child to be not future date. Try to think about correct solution for your  @PastOrPresent annotation

kostyaskrun commented on src/main/java/com/example/bookstore/model/Book.java:
```diff
@@ -0,0 +1,75 @@
+    private String title;
+
+    @Column(name = "published_date")
+    @PastOrPresent(message = "Published date cannot be in the future")
```
ah, I thought you created this annotation by yourself, but jakarta.validation also works fine, nice

kostyaskrun reviewed (changes requested):
Please add integration test, check my previous comment

sher009 committed:
Validation to Author and Book

kostyaskrun commented on src/main/java/com/example/bookstore/model/Book.java:
```diff
@@ -0,0 +1,75 @@
+
+
+import jakarta.persistence.*;
+import jakarta.validation.constraints.PastOrPresent;
```
you are still using jakarta.validation.constraints.PastOrPresent not com.example.bookstore.validation.PastOrPresent - fix this

kostyaskrun commented on src/main/java/com/example/bookstore/model/Author.java:
```diff
@@ -1,8 +1,11 @@
 
 import jakarta.persistence.*;
+import jakarta.validation.constraints.NotNull;
+import jakarta.validation.constraints.Past;
```
you are still using jakarta.validation.constraints.Past not com.example.bookstore.validation.PastOrPresent - fix this

sher009 replied:
NotNull I left

kostyaskrun commented on src/main/java/com/example/bookstore/model/Author.java:
```diff
@@ -28,6 +34,9 @@ public class Author {
     @Column(name = "language")
     private String language;
 
+    @OneToMany(mappedBy = "author", cascade = CascadeType.ALL, orphanRemoval = true)
```
Please create integration test to see what happens with Books when Author is deleted

kostyaskrun commented on src/test/java/com/example/bookstore/BookValidationTest.java:
```diff
@@ -0,0 +1,35 @@
+
+import static org.junit.jupiter.api.Assertions.assertFalse;
+
+public class BookValidationTest {
```
there is should be only one test covering your custom @PastOrPresent annotation,  AuthorValidationTest and BookValidationTest just a duplicate of same code. Fix this

kostyaskrun replied:
hint: test annotation logic (PastOrPresentValidator) not class which uses annotation (Author, Book)

kostyaskrun commented on src/main/java/com/example/bookstore/BookService.java:
```diff
@@ -0,0 +1,32 @@
+        this.bookRepository = bookRepository;
+    }
+
+    public List<Book> getAllBooks() {
```
nit:

no need to specify "book" in all method names, it's already clear from the class name that we work here with Books, 

for now it's fine to have such method namings:

"getAll"
"getById"
"save"

etc

kostyaskrun commented on src/test/java/service/BookServiceTest.java:
```diff
@@ -0,0 +1,54 @@
+import static org.mockito.ArgumentMatchers.any;
+import static org.mockito.Mockito.*;
+
+public class BookServiceTest {
```
delete and getAll methods missed in test

sher009 committed:
1.connect to own annotation class Book and Author
2.test annotation logic (PastOrPresentValidator) not class which uses annotation (Author, Book)
3.deleted(no need to specify "book")
4.add test delete and getAll methods

kostyaskrun commented on src/main/java/com/example/bookstore/controller/AuthorController.java:
```diff
@@ -35,6 +37,15 @@ public Author createAuthor(@RequestBody Author author) {
         return authorService.saveAuthor(author);
     }
 
+    @PostMapping
```
you have duplicate of @PostMapping createAuthor

kostyaskrun commented on src/main/java/com/example/bookstore/controller/BookController.java:
```diff
@@ -0,0 +1,55 @@
+    }
+
+    @PostMapping
+    public ResponseEntity<?> createBook(@Valid @RequestBody Book book, BindingResult result) {
```
why do we need two createBook methods?

kostyaskrun commented on src/test/java/com/example/bookstore/AuthorValidationTest.java:
```diff
@@ -0,0 +1,35 @@
+
+import static org.junit.jupiter.api.Assertions.assertFalse;
+
+public class AuthorValidationTest {
```
most probably, you forgot to delete this test class

kostyaskrun commented on src/test/java/com/example/bookstore/BookValidationTest.java:
```diff
@@ -0,0 +1,35 @@
+
+import static org.junit.jupiter.api.Assertions.assertFalse;
+
+public class BookValidationTest {
```
most probably, you forgot to delete this test class

kostyaskrun commented on src/test/java/com/example/bookstore/PastOrPresentValidatorTest.java:
```diff
@@ -0,0 +1,60 @@
+import static org.junit.jupiter.api.Assertions.assertFalse;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+
+public class PastOrPresentValidatorTest {
```
finally, perfect solution. nice job

sher009 replied:
thank you

kostyaskrun commented on src/test/java/com/example/bookstore/PastOrPresentValidatorTest.java:
```diff
@@ -0,0 +1,60 @@
+        PastOrPresentValidator validator = new PastOrPresentValidator();
+
+        //Validate the past date (should pass)
+        assertTrue(validator.isValid(pastDate, null), "Date should not be in the past");
```
```suggestion
        assertTrue(validator.isValid(pastDate, null), "Date should be in the past");
```

kostyaskrun commented on src/test/java/com/example/bookstore/PastOrPresentValidatorTest.java:
```diff
@@ -0,0 +1,60 @@
+        PastOrPresentValidator validator = new PastOrPresentValidator();
+
+        //Validate null value (should pass)
+        assertTrue(validator.isValid(null, null), "Null Date should not be in the past");
```
```suggestion
        assertTrue(validator.isValid(null, null), "Null Date allowed");
```

sher009 committed:
1.Fixed duplicate of @PostMapping in BookController and AuthorController
2.Deleted AuthorValidationTest and BookValidationTest
3.Fixed comments
